x-pack/plugins/monitoring_collection/server/routes/dynamic_route.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { Observable } from 'rxjs';
+import { JsonObject } from '@kbn/utility-types';
+import { schema } from '@kbn/config-schema';
+import { IRouter, ServiceStatus } from '../../../../../src/core/server';
+import { getESClusterUuid, getKibanaStats } from '../lib';
+import { MetricResult } from '../plugin';
+
+export function registerDynamicRoute({
+  router,
+  config,
+  overallStatus$,
+  getMetric,
+}: {
+  router: IRouter;
+  config: {
+    allowAnonymous: boolean;
+    kibanaIndex: string;
+    kibanaVersion: string;
+    uuid: string;
+    server: {
+      name: string;
+      hostname: string;
+      port: number;
+    };
+  };
+  overallStatus$: Observable<ServiceStatus>;
+  getMetric: (
+    type: string
+  ) => Promise<Array<MetricResult<JsonObject>> | MetricResult<JsonObject> | undefined>;
+}) {
+  router.get(
+    {
+      path: `/api/monitoring_collection/{type}`,
+      options: {
+        authRequired: !config.allowAnonymous,
+        tags: ['api'], // ensures that unauthenticated calls receive a 401 rather than a 302 redirect to login page
+      },
+      validate: {
+        params: schema.object({
+          type: schema.string(),
+        }),
+      },
+    },
+    async (context, req, res) => {
+      const type = req.params.type;
+      const [data, clusterUuid, kibana] = await Promise.all([
+        getMetric(type),
+        getESClusterUuid(context.core.elasticsearch.client),
```

Comment:
How often is the cluster UUID really likely to change? I honestly don't know, but this doesn't seem like something that needs to be queried on every API request. Could it be cached in internally and periodically refreshed, if needed?

Reply:
Yea definitely a good idea. I wonder how often it could change - I don't think it's safe to assume we can query it once and store the value in memory indefinitely. Do we think it's a problem as it will only be calculated when the API is called?

Reply:
Today, I think it doesn't change. I think Kibana stores a lot of state about the user's cluster, that if the UUID were to switch it would disrupt a lot of internal state. But I'm starting to think that making Kibana keep internal state about ES, or any upstream service, is a bad idea. I withdraw my question for now.

Reply:
I think this value changes a lot in Cloud, FWIW.